compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

```diff
@@ -533,7 +533,7 @@ class SpaceEngine(using Context) extends SpaceLogic {
   /** Is `tp1` a subtype of `tp2`?  */
   def isSubType(tp1: Type, tp2: Type): Boolean = trace(i"$tp1 <:< $tp2", debug, show = true) {
     if tp1 == constantNullType && !ctx.explicitNulls then tp2 == constantNullType
-    else adaptType(tp1, tp2) <:< tp2
+    else tp1 <:< tp2 || adaptType(tp1, tp2) <:< tp2
```

Comment:
Just curious: I'm wondering if it is possible to push `<:< tp2` into `adaptType(tp1, tp2)` to make the new "subtyping" rule more explicit? (I haven't thought through the details, please ignore if it does not make sense technically).

Reply:
I can't tell if I'm being cautious or just short-sighted and bone-headed, but I kind of don't want to get into the business of trying to integrate the type-fudging we're doing here, tightly into the subtyping rules.  Re-reading your earlier "augment the rules" comment, realising it's what you meant, rather than this.  We tried an earlier suggestion from Guillaume of using "isPrimToBox" alone, but that only considers the classSymbol alone, incorrectly making `Integer <:< (1: Int)` true.  So maybe the "augment" idea isn't a good idea because the subtyping rules are already so complex?  Or, again, I'm being short-sighted.

In many ways adapting types has a lot of precedent in the codebase, such as in typer's tree adapting (where I lifted the ConstantType converting code) but also in TypeOps where there's a lot of type mapping to approximate things and try subtyping multiple times.

In many ways the subtyping rules are fixed, but I get the sense that even though fundamental changes don't happen often, there are changes in encoding or implementation detail that put all these auxiliary type mapping and approximating code paths at risk of diverging from their original intent (when unguarded by tests).

WDYT?  Am I being too paranoid?  Do you think it's reasonable easy for me to just sit down for a second and figure out how to get these silly cases to give the right answer?

Reply:
> I kind of don't want to get into the business of trying to integrate the type-fudging we're doing here, tightly into the subtyping rules

Sorry I wasn't clear. I don't mean change `<:<`. What I was proposing is some refactoring. Instead of `tp1 <:< tp2 || adapt(tp1, tp2) <:< tp2`, what about `tp1 <:< tp2 || primitiveSubtype(tp1, tp2)`?

As I mentioned, this is a minor thing, more of a style --- the current PR is good as it is.

Reply:
Oh I had understood that.  By integrating I meant `primitiveSubtype` must be some subset of the subtyping rules, which at least handles that `Integer <:< (1: Int)` is false.